Add Typed Step Builder creation pattern (#105)
ponylang/pony-patterns: docs/creation/typed-step-builder.md

```

This compiles and runs, producing a message with no recipient and no subject. The builder silently accepted incomplete data because every field had a default. The caller forgot `.to()` and the compiler didn't say a word.

You could add validation to `build()` — check that `_to` isn't empty and return an error or raise `error` if it is. That catches the mistake, but now the caller has to handle a runtime error every time they build a message, even when they did fill in every field. The compiler can't tell a correct build chain from a broken one, so every call site pays the cost of error handling.

## Solution

Instead of a single builder type with all the methods, define a separate interface for each construction phase. Each interface exposes exactly one advancement method whose return type is the next phase's interface. The compiler enforces the build order: you literally can't call the wrong method because it doesn't exist on the type you're holding.

We'll build an email message in three mandatory steps: set the recipient, set the subject, then set the body. Here's the first phase:

```pony
interface MessageRecipient
  fun ref to(recipient: String): MessageSubject
```

> A `MessageRecipient` has exactly one method: `to()`. It takes the recipient and returns a `MessageSubject`. There's no way to finish without providing a recipient first.

Reviewer: ```suggestion
A `MessageBuildRecipient` has exactly one method: `to()`. It takes the recipient and returns a `MessageBuildSubject`. There's no way to finish without providing a recipient first.
```

```pony
interface MessageSubject
  fun ref subject(subject': String): MessageBody
```

Once you've set the recipient, you're holding a `MessageSubject`. The only thing you can do is call `subject()`, which advances you to the final phase.

```pony

> interface MessageBody

Reviewer: ```suggestion
interface MessageBuildBody
```

Now the concrete class that ties it all together:

```pony
class _MessageBuilder
  var _to: String = ""
  var _subject: String = ""

  fun ref to(recipient: String): MessageSubject =>
    _to = recipient
    this

  fun ref subject(subject': String): MessageBody =>
    _subject = subject'
    this

  fun ref body(body': String): String =>
    "To: " + _to + "\nSubject: " + _subject + "\n\n" + body'
```

`_MessageBuilder` has all three methods, one from each phase. Notice there's no `is` clause declaring that it implements any of the interfaces. It doesn't need one. Pony interfaces use structural subtyping: any type whose methods match an interface's signatures automatically satisfies that interface. Since `_MessageBuilder` has `to()` returning `MessageSubject`, it satisfies `MessageRecipient`. Since it has `subject()` returning `MessageBody`, it satisfies `MessageSubject`. And since it has `body()` returning `String`, it satisfies `MessageBody`. One class, three interface views.

The caller never sees `_MessageBuilder` directly. A factory primitive provides the entry point:

```pony
primitive Messages

>   fun apply(): MessageRecipient =>

Reviewer: ```suggestion
  fun apply(): MessageBuildRecipient =>
```

    _MessageBuilder
```

`Messages()` returns a `MessageRecipient`, so the caller starts at phase one. Each method advances to the next phase, and the compiler enforces the order.

Here's the complete program:

```pony
interface MessageRecipient

>   fun ref to(recipient: String): MessageSubject

Reviewer: ```suggestion
  fun ref to(recipient: String): MessageBuildSubject
```

> interface MessageSubject

Reviewer: ```suggestion
interface MessageBuildSubject
```

  fun ref subject(subject': String): MessageBody

> interface MessageBody

Reviewer: ```suggestion
interface MessageBuildBody
```

  fun ref body(body': String): String

class _MessageBuilder
  var _to: String = ""
  var _subject: String = ""

  fun ref to(recipient: String): MessageSubject =>
    _to = recipient
    this

  fun ref subject(subject': String): MessageBody =>
    _subject = subject'
    this

  fun ref body(body': String): String =>
    "To: " + _to + "\nSubject: " + _subject + "\n\n" + body'

primitive Messages

>   fun apply(): MessageRecipient =>

Reviewer: ```suggestion
  fun apply(): MessageBuildRecipient =>
```

The pattern extends naturally to repeatable steps within a phase. If one of your phases allows multiple calls before advancing, give it a method that stays in the current phase alongside the advancement method. For example, a `cc()` method on `MessageRecipient` could return `MessageRecipient` while `to()` still advances to `MessageSubject`. The caller can add as many CC recipients as they like, and the compiler still forces them to call `to()` before moving on. The same idea works for optional fields: put them on the phase interface alongside the required advancement method.

The Typed Step Builder becomes especially powerful when combined with constrained types. If `to()` accepts an `EmailAddress` that can only be constructed through validation instead of a bare `String`, the builder enforces not just that a recipient was provided but that it's a well-formed address. The builder handles ordering at compile time; constrained types handle value validity at the point of construction. Together they guarantee the final object has all required fields, provided in the right order, with every field individually validated.

> Builders can also support reuse. Adding a `reset()` method to each interface that returns `MessageRecipient` gives the caller a way to abandon a partial build and start over from the first phase without allocating a new builder.

Reviewer: ```suggestion
Builders can also support reuse. Adding a `reset()` method to each interface that returns `MessageBuildRecipient` gives the caller a way to abandon a partial build and start over from the first phase without allocating a new builder.
```
